### PR TITLE
Option to generate usage report from a celery task

### PR DESCRIPTION
**Pull request by marcospri**
Store the results in the DB for:

- Retrieval of the report without having to do the query work
- Option to graph the results in metabase over a period of time.

This commit add the option to generate these in an async (celery) task
but not code uses it yet.

The plan is to write a second  task that schedules periodic reports for
ongoing company deals.

## Deploy

- Remember to run migration in production

## Testing

- Apply migration

```
tox -e dev --run-command 'alembic upgrade head'
```

- In `make shell` kick off a job

```
from datetime import datetime, timedelta
from lms.tasks.organization import generate_usage_report

org_public_id = "us.lms.org.LDtcl7EUTeW2UERPJLAVtA"

generate_usage_report.delay(db.query(models.Organization).filter_by(public_id=org_public_id).one().id, "testing", (datetime.now() - timedelta(days=365)).isoformat(), datetime.now().date().isoformat())
```

- Check the job finished, in the `make dev` logs

```
Task lms.tasks.organization.generate_usage_report[1b603b4c-7f06-45ac-9429-d84f8d00873b] received
lms.tasks.organization.generate_usage_report[1b603b4c-7f06-45ac-9429-d84f8d00873b] Task lms.tasks.organization.generate_usage_report[1b603b4c-7f06-45ac-9429-d84f8d00873b] succeeded in 0.20881368702976033s: None
```

- And finally check the new reports stored in the DB, in the same `make shell`

```
db.query(models.OrganizationUsageReport).first().__dict__


{'_sa_instance_state': <sqlalchemy.orm.state.InstanceState at 0x7595e24c0860>, 'since': datetime.date(2023, 6, 5),
 'created': datetime.datetime(2024, 6, 4, 7, 53, 4, 669763), 'updated': datetime.datetime(2024, 6, 4, 7, 53, 4, 669763),
 'until': datetime.date(2023, 6, 5), 'tag': 'testing',
 'organization_id': 1, 'key': '1-testing-2023-06-05 09:53:04.513550-2024-06-04 00:00:00',
 'report': [{'name': 'DISPLAY NAME TEACHER',   'email': 'eng+canvasteacher@hypothes.is',
   'h_userid': 'acct:3a022b6c146dfd9df4ea8662178eac@lms.hypothes.is',   'course_name': 'Developer Test Course with Sections Enabled',
   'course_created': '2024-04-18',   'authority_provided_id': '653e9620a656a954c684942f6443fa3e3410c03a'},
  {'name': 'DISPLAY NAME TEACHER',   'email': 'eng+canvasteacher@hypothes.is',
   'h_userid': 'acct:3a022b6c146dfd9df4ea8662178eac@lms.hypothes.is',   'course_name': 'LTI 1.3 Testing',
   'course_created': '2024-04-18',   'authority_provided_id': '3394ac098caa5997ed3d87023befb9f7e9ea4a38'},
  {'name': '<STUDENT>',   'email': '<STUDENT>',
   'h_userid': 'acct:f6c3ed0edd8a4013bfc8d0c22b9eca@lms.hypothes.is',   'course_name': 'Developer Test Course with Sections Enabled',
   'course_created': '2024-04-18',   'authority_provided_id': '653e9620a656a954c684942f6443fa3e3410c03a'},
  {'name': '<STUDENT>',   'email': '<STUDENT>',
   'h_userid': 'acct:f6c3ed0edd8a4013bfc8d0c22b9eca@lms.hypothes.is',   'course_name': 'LTI 1.3 Testing',
   'course_created': '2024-04-18',   'authority_provided_id': '3394ac098caa5997ed3d87023befb9f7e9ea4a38'},
  {'name': 'Michael Admin',   'email': 'mdiroberts+canvasadmin@hypothes.is',
   'h_userid': 'acct:8652fe5b4ab4820743c9b7640f3aea@lms.hypothes.is',   'course_name': 'Developer Test Course with Sections Enabled',
   'course_created': '2024-04-18',   'authority_provided_id': '653e9620a656a954c684942f6443fa3e3410c03a'},
  {'name': '<STUDENT>',   'email': '<STUDENT>',
   'h_userid': 'acct:c2cd2d06fc04ff4f0799e8e17c8c27@lms.hypothes.is',   'course_name': 'LTI 1.3 Testing',
   'course_created': '2024-04-18',   'authority_provided_id': '3394ac098caa5997ed3d87023befb9f7e9ea4a38'}],
 'id': 1, 'unique_users': 4}
```

### CODE REVIEW

File: lms/services/organization.py [hunk 1]
```diff
@@ -316,7 +351,7 @@ def usage_report(
                 email=row.email if row.email else "<STUDENT>",
                 h_userid=row.h_userid,
                 course_name=row.course_name,
-                course_created=row.course_created,
+                course_created=row.course_created.isoformat(),
```
Comment (marcospri) [inline lms/services/organization.py:354]: This is json serializable, the other usage of this reports is dumping it on the admin pages so this change doesn't require any other adjustments.

File: lms/services/organization.py [hunk 2]
```diff
@@ -238,6 +239,40 @@ def update_organization(  # noqa: PLR0913
...
+        report = OrganizationUsageReport(
+            organization=organization,
+            key=report_key,
+            since=since,
+            until=since,
+            tag=tag,
+        )
+        self._db_session.add(report)
+
+        report_rows = self.usage_report(organization, since, until)
```
Comment (marcospri) [inline lms/services/organization.py:269]: Use the existing method to get the report with the right parameters.

File: lms/services/organization.py [hunk 3]
```diff
@@ -238,6 +239,40 @@ def update_organization(  # noqa: PLR0913
...
+
+        report.unique_users = len({r.h_userid for r in report_rows})
```
Comment (marcospri) [inline lms/services/organization.py:271]: Store the count of unique users ( { } is a set)

File: lms/services/organization.py [hunk 4]
```diff
@@ -238,6 +239,40 @@ def update_organization(  # noqa: PLR0913
...
+        report.report = [asdict(r) for r in report_rows]
```
Comment (marcospri) [inline lms/services/organization.py:272]: And the full report.

File: lms/tasks/organization.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+from datetime import datetime
+
+from lms.services import OrganizationService
+from lms.tasks.celery import app
+
+
+@app.task
+def generate_usage_report(
+    organization_id: int, tag: str, since: str, until: str
```
Comment (marcospri) [inline lms/tasks/organization.py:9]: Nothing fancy here, just take the params needed as json serializable parameters and call the service method.

File: lms/tasks/organization.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+) -> None:
+    with app.request_context() as request:
+        with request.tm:
+            request.find_service(OrganizationService).generate_usage_report(
+                organization_id,
+                tag,
+                datetime.fromisoformat(since),
```
Comment (marcospri) [inline lms/tasks/organization.py:16]: Use datetime in the service but strings here (to be able to use celery json serialization)

File: lms/services/organization.py [hunk 7]
```diff
@@ -238,6 +239,40 @@ def update_organization(  # noqa: PLR0913
 
         return organization
 
+    def generate_usage_report(
+        self, organization_id: int, tag: str, since: datetime, until: datetime
+    ):
+        """Generate and store an usage report for one organization."""
```
Comment (acelaya) [inline lms/services/organization.py:245]:
```suggestion
        """Generate and store a usage report for one organization."""
```

File: lms/services/organization.py [hunk 8]
```diff
@@ -238,6 +239,40 @@ def update_organization(  # noqa: PLR0913
...
+            .one_or_none()
+        ):
+            LOG.debug("Report already exists, skipping generation. %s", report_key)
+            return report
+
+        report = OrganizationUsageReport(
+            organization=organization,
+            key=report_key,
+            since=since,
+            until=since,
```
Comment (acelaya) [inline lms/services/organization.py:264]: Is this intended, or is it a typo?

```suggestion
            until=until,
```

**Review by acelaya — APPROVED** [review]
LGTM. Just found a potential typo when creating an `OrganizationUsageReport` instance.

File: lms/services/organization.py (hunk 8 above)
Comment (marcospri) [reply lms/services/organization.py]: Yes!, good catch. Added the missing test assertion for this.

**Pull request merged**